### DISCUSSION TITLE
Submitting a tx to your own node with cardano-submit-api

### DISCUSSION THREAD

**Discussion by astrojarred** [Python-Cardano/pycardano#13] [Show and tell]
[Cardano Submit API ](https://github.com/input-output-hk/cardano-node/tree/master/cardano-submit-api) is packaged within the Cardano Node repository and lets you submit transactions via a REST API endpoint.

I don't know if it makes sense to include this functionality in pycardano since Ogmios will cover most people's use-cases, but with all the network congestion going on I've been playing around with using BlockFrost for all queries but submitting via my own node.

It's pretty straightforward to submit a tx this way if you have `cardano-submit-api` already up and running, but I just wanted to document it here in case it might be useful for others.

### Submitting a `pycardano` transaction to a local node with `cardano-submit-api`:
```python
import requests

# ... build your tx ending with a final signed Tx object
signed_tx = pycardano.Transaction(tx_body, pycardano.TransactionWitnessSet(vkey_witnesses=vk_witnesses))

# submit tx via cardano-submit-api
submit_api_url = "http://{yourURLorLocalhost}/api/submit/tx"

data = bytes.fromhex(signed_tx.to_cbor())  # convert CBOR to bytes

headers = {'Content-type': 'application/cbor'}  # create headers

response = requests.post(submit_api_url, data=data, headers=headers)

if not response.ok:
    print("Issue submitting tx:")
    print(response.json())
else:
    tx_id = response.json()  # you can also get the outgoing Tx ID directly from signed_tx with `str(signed_tx.id)`

    print(f"Successfully submitted {tx_id}")
```

**Comment by cffls** [comment]
Thanks for providing this idea! Agree that BlockFrost is probably not the best solution for submitting transactions (hope they can solve this problem soon). Also, it is possible to implement a ChainContext with cardano-node/cardano-cli if we see values in the future.

**Comment by astrojarred** [comment]
> Also, it is possible to implement a ChainContext with cardano-node/cardano-cli if we see values in the future.

Was thinking about this a bit too. I assume it would just be a lot of wrapping CLI commands and running them with subprocess. If there is ever a demand for it I could be willing to build this out, but I think for most people the most convenient method is probably just to use BlockFrost for everything but tx submissions.

**Reply by cffls** [reply]
Yeah, it is definitely going to be a lot of works, but thanks for the offer! :D  Another solution is to use Ogmios chain context so we can get faster transaction submission. Although that means we need to run a local node and Ogmios service, it is very simple with a docker compose file.

**Reply by reitcircles** [reply]
Would you know how to implement ogmios context instead of blockfrost. ? Any examples would help. thanks
